src/circuit/gadgets/constraints.rs

```diff
@@ -806,11 +805,28 @@ mod tests {
 
     use bellperson::util_cs::test_cs::TestConstraintSystem;
     use blstrs::Scalar as Fr;
+    use ff::Field;
     use proptest::prelude::*;
     use std::ops::{AddAssign, SubAssign};
 
     use crate::field::FWrap;
 
+    // Auxiliary function for creating a random linear combination
+    fn random_lc_strategy(cs: &mut impl ConstraintSystem<Fr>, n: usize) -> LinearCombination<Fr> {
+        let mut coeffs = vec![0; n];
+        coeffs.fill_with(|| rand::thread_rng().gen_range(0..10));
+
+        let mut lc = LinearCombination::zero();
+        for (i, coeff) in coeffs.iter().enumerate() {
+            let allocated_var =
+                AllocatedNum::alloc(cs.namespace(|| format!("{i}")), || Ok(Fr::zero())).unwrap();
+            let var = allocated_var.get_variable();
+            let val = Fr::from_u64(*coeff);
+            lc = lc + (val, var);
```

Comment:
After the latest updates it's zero a fifth of the time only.

Reply:
I don't think so. I think each component is 0 one fifth of the time, and 1 one fifth of the time too. But their sum will actually be 0 only (1/5)^5 = 1/3125 of the time (when all components are 0). That's 'fine', since 0 is not an important value. However, the sum of the components will only be 1 (which *is* significant) when exactly one is 1. That means of the 3125 possibilities, only five will yield a sum of 1  — so once in 625 times.

This means that in terms of my previous suggestion to either just enumerate boolean lcs or some *small* number of them, we're instead enumerating 3125 values, of which 625 are 0 and 5 are 1. This still seems wasteful. We'll need to end up doing more than a thousand iterations to make sure we probabilistically exercise the 'normal case' where premise is true and lc==1. Since we only need to exercise 4 cases (or 6, or 8, or 10, or whatever), this seems wasteful and accomplished in a needlessly elaborate way.

Reply:
Ah, sorry @huitseeker — although my comments above stand with regard to the sum, I see that you also added a simpler test which I missed in only reading your response here. Sorry for missing that this isn't meant to do the same job as that anymore.


Still, having looked, I remain unconvinced this lc strategy does what you want *for this specific case* (as above). That said, I do see the value of having a random lc strategy in general. Maybe what we need is to more clearly define the goals/spec and perhaps add some parameters. The danger is that since such bits are inherently very tricky, as we've seen, once this becomes reified as a 'thing', we'll tend to use it without too much thought. So we do need to document very clearly what the expected behavior is.

The perhaps dubious insight I am offering is that where a linear combination is needed in tests, it probably needs to be a linear combination of some other, independently-significant-to-the-circuit values (of whatever type). Otherwise, the exercise of assembling a linear combination from components doesn't add anything. We might as well sample a random number from the distribution we care about and create a simple linear combination of that value times one. This would make the intent much clearer and reasoning about composition in future proptests less error-prone.